# Adatok betöltése

In [107]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

In [108]:

import types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share the notebook.
client_bc050f45180b4c1bb8886542b85a43fe = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='PJI3OXoebDw8yJ2ju12loFQ9I1R6ljjP_fSBbeRnBfaE',
    ibm_auth_endpoint="https://iam.cloud.ibm.com/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3-api.us-geo.objectstorage.service.networklayer.com')

body = client_bc050f45180b4c1bb8886542b85a43fe.get_object(Bucket='stratis-donotdelete-pr-uyyguvdau1klvi',Key='texts.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_1 = pd.read_csv(body)
df_data_1


Unnamed: 0    Topic                                               Text
0              0  techtud  A Magyar Tudományos Akadémia (MTA) elnöksége n...
1              1  techtud  Azt eddig is tudni lehetett, hogy a hőmérsékle...
2              2  techtud  Vlagyimir Putyin orosz elnök aláírta azt az áp...
3              3  techtud  A marsi napfelkeltéről és a napnyugtáról küldö...
4              4  techtud  Egy új kísérlet szerint a farkasok hajlamosabb...
5              5  techtud  Franciaország legnagyobb akváriuma egy meglehe...
6              6  techtud  Lézeres technológiát használva kutatók új maja...
7              7  techtud  A Valve bemutatta az Index nevű VR-kütyükészle...
8              8  techtud  Az angol kutatók a mikroszennyezést vizsgálták...
9              9  techtud  A Boeing két nagy repülőgép-balesetben is hibá...
10            10  techtud  Karantén alá vonták Mongólia Bajan-Ölgij tarto...
11            11  techtud  Egy dél-koreai azt állítja, hogy minden különö...
12            12  techtud  Múlt hétvégén, péntek éjjel 11 és szombat hajn...
13            13  techtud  Nincsenek könnyű helyzetben most a Leica fejes...
14            14  techtud  San Joséban tartja a Facebook a fejlesztői kon...
15            15  techtud  A cseh koalíciós kormány megállapodásának érte...
16            16  techtud  Brit kutatók szerint a vizeletvizsgálat ugyano...
17            17  techtud  Most először találták meg a csillagászok annak...
18            18  techtud  Tízszer gyorsabban olvad a világ legnagyobbjak...
19            19  techtud  Érzékeny adatokat tartalmazó védtelen szervert...
20            20  techtud  Az emberek nagyon nagy része kifejezetten utál...
21            21  techtud  A GDPR, azaz az Európai Unió általános adatvéd...
22            22  techtud  A polimerre általában szigetelőként szokott go...
23            23  techtud  Több mint száz éve próbálják már megfejteni a ...
24            24  techtud  Jonah M. Kessel és kollégái évek óta gyűjtenek...
25            25  techtud  Amikor új csúcstelefont dob egy cég piacra, mi...
26            26  techtud  A talaj károsítása gyorsíthatja a klímaváltozá...
27            27  techtud  Megújult a Magyar Telekom mobilos kínálata, az...
28            28  techtud  A Samsung több új tévémodellt is bemutatott a ...
29            29  techtud  A szülők közmondásos elfogultsága a gyerekek e...
...          ...      ...                                                ...
1273        1290    sport  Hiába tért vissza Zinedine Zidane a csapathoz,...
1274        1291    sport  Hiába kezdett jól Babos Tímea a szerb Ivana Jo...
1275        1292    sport  Hosszabb pihenőt tervez az NBA-szezon végét sé...
1276        1293    sport  Csapatsorrendet akart alkalmazni a vasárnapi, ...
1277        1294    sport  Vettel beletörődött a Merci-fölénybe, Leclerc-...
1278        1295    sport  Még január végén jött a hír, hogy az átigazolá...
1279        1296    sport  \nAz Arsenal nagyon simán, 2-0-ra győzte le a ...
1280        1297    sport  Mesés rajtpénzt kapott a 2009-es ausztrál körv...
1281        1298    sport  \nAz Everton vizsgálja, hogy mi is történhetet...
1282        1299    sport  Ole Gunnar Solskjaer szerint a világbajnok Pau...
1283        1300    sport  \nBelgium 4-2-re verte a magyar válogatottat, ...
1284        1301    sport  Seszták Miklós futballbirodalmában gordiuszi m...
1285        1302    sport  \nNem sikerült a bravúr, Stollár Fanny két sze...
1286        1303    sport  Matteo Salvini olasz belügyminiszter kijelente...
1287        1304    sport  Múlt héten tartották Barcelonában a WTCR szezo...
1288        1305    sport  15 győztes meccs után a 16. találkozóján is es...
1289        1306    sport  Kínosan nagy lesgólt kapott meg a Chelsea a Ca...
1290        1307    sport  Másfél éves előkészítő munka után a jövő héten...
1291        1308    sport  A nemzetközi sajtó alig ír a győztes Lewis Ham...
1292        1309    sport  A MagyarKupa hétvégi négyes döntője után nem s...
12

In [109]:
df_data_1.shape

(1303, 3)

A sorszám oszlop néha ugrik egyet (pl 130 kimarad) ennek köszönhető hogy 1302 mellett 1319 van, a megfigyelések valós száma 1303. Erre az oszlopra amúgy sincs szükség törlöm.

In [110]:
topic_rec=df_data_1.drop(['Unnamed: 0'], axis=1)
topic_rec.head()

Topic                                               Text
0  techtud  A Magyar Tudományos Akadémia (MTA) elnöksége n...
1  techtud  Azt eddig is tudni lehetett, hogy a hőmérsékle...
2  techtud  Vlagyimir Putyin orosz elnök aláírta azt az áp...
3  techtud  A marsi napfelkeltéről és a napnyugtáról küldö...
4  techtud  Egy új kísérlet szerint a farkasok hajlamosabb...

In [111]:
X1=topic_rec[['Text']]
X1=X1['Text'].tolist()

#### Bag of words módszerrel a cikkek átfordítása 0/1 számokra

In [112]:
import keras
from keras.preprocessing.text import Tokenizer
tokenizer=Tokenizer(num_words=3000,lower=True)
tokenizer.fit_on_texts(X1)

In [113]:
X=tokenizer.texts_to_matrix(X1, mode='binary')
X.shape

(1303, 3000)

#### A szöveges célváltozók cseréje számokra

In [114]:
topic_rec=topic_rec.replace('techtud', 0)
topic_rec=topic_rec.replace('kultur', 1)
topic_rec=topic_rec.replace('sport', 2)
topic_rec.head()

Topic                                               Text
0      0  A Magyar Tudományos Akadémia (MTA) elnöksége n...
1      0  Azt eddig is tudni lehetett, hogy a hőmérsékle...
2      0  Vlagyimir Putyin orosz elnök aláírta azt az áp...
3      0  A marsi napfelkeltéről és a napnyugtáról küldö...
4      0  Egy új kísérlet szerint a farkasok hajlamosabb...

In [115]:
topic_rec.Topic.unique()

array([0, 1, 2])

In [116]:
Y=topic_rec[['Topic']].values
Y.shape

(1303, 1)

#### Train és Test felbontás

In [117]:
from sklearn.model_selection import train_test_split
X_trainset, X_testset, Y_trainset, Y_testset = train_test_split(X, Y, test_size=0.3, random_state=3)
X_trainset.shape

(912, 3000)

In [118]:
Y_trainset.shape

(912, 1)

#### Modellezés 

In [119]:
from keras.utils import to_categorical


model = keras.Sequential([
    keras.layers.Dense(32, input_shape=(3000,)  ,activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(32, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(3, activation='softmax')
    
])
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

In [124]:
model.fit(X_trainset, to_categorical(Y_trainset), epochs=20)

Epoch 1/20
912/912 [==============================] - 1s 616us/step - loss: 0.0189 - acc: 0.9967
Epoch 2/20
912/912 [==============================] - 1s 615us/step - loss: 0.0201 - acc: 0.9923
Epoch 3/20
912/912 [==============================] - 1s 1ms/step - loss: 0.0264 - acc: 0.9923
Epoch 4/20
912/912 [==============================] - 1s 643us/step - loss: 0.0200 - acc: 0.9956
Epoch 5/20
912/912 [==============================] - 1s 652us/step - loss: 0.0203 - acc: 0.9934
Epoch 6/20
912/912 [==============================] - 1s 696us/step - loss: 0.0218 - acc: 0.9956
Epoch 7/20
912/912 [==============================] - 1s 815us/step - loss: 0.0180 - acc: 0.9956
Epoch 8/20
912/912 [==============================] - 1s 652us/step - loss: 0.0131 - acc: 0.9989
Epoch 9/20
912/912 [==============================] - 1s 640us/step - loss: 0.0156 - acc: 0.9967
Epoch 10/20
912/912 [==============================] - 1s 634us/step - loss: 0.0132 - acc: 0.9978
Epoch 11/20
912/912 [==========

#### Modell kiértékelése

In [125]:
model.evaluate(
  X_testset,
  to_categorical(Y_testset)
)

391/391 [==============================] - 0s 214us/step


[0.32321343285119747, 0.9386189258312021]

#### Paraméterek amit tune-oltam, hogy minél jobb eredmény legyen teszt mintán:
- n -> különböző  leggyakoribb szavak száma ami jön a bag of words-ből
- neurális háló struktúrája -> kell-e penalizaton/hány neuron legyen
- epoch szám

#### Többszöri próbálkozás eredménye összegezve:
- 3000-nél sokkal alacsonyabb n jelentősen ront a modell teljesítményén, de nagyobb n nem tud javítani
- penalization javít a modell teljesítményén ez Dropout-tal lett megoldva
- 32 neuron ideális egy layerben -> 16-al még gyengébb a teljesítmény, 64-nél már nem javul
- epoch szám majdnem mindegy ha elég magas, mert az early stopping-ot mint overfit megakadályozása most nem használtam
- a legjobb teszt minta eredmények mind 92-94% körül voltak
- 99,5% a train mintán a teljesítmény, amitől gyanús, hogy overfit a modell, de kevésbé overfit modellek (96, 97% pontosságúak train mintán) sem értek el jobb teljesítményt teszt mintán 


#### Modell becslés vs. input

In [126]:
A=model.predict(X_testset)
A[0:10]

array([[4.5169452e-14, 1.3933851e-14, 1.0000000e+00],
       [3.1656012e-24, 2.8487841e-26, 1.0000000e+00],
       [5.2696848e-18, 8.1748911e-21, 1.0000000e+00],
       [9.9994946e-01, 4.9926493e-05, 5.6312518e-07],
       [2.1121194e-08, 5.0022547e-10, 1.0000000e+00],
       [2.3924073e-07, 1.7084863e-08, 9.9999976e-01],
       [1.8093351e-05, 9.9998057e-01, 1.3503824e-06],
       [4.9481417e-07, 4.8162235e-08, 9.9999940e-01],
       [5.5063487e-07, 4.9544948e-08, 9.9999940e-01],
       [7.1022164e-08, 9.9999988e-01, 2.2182850e-10]], dtype=float32)

In [127]:
to_categorical(Y_testset)[0:10]

array([[0., 0., 1.],
       [0., 0., 1.],
       [0., 0., 1.],
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 0., 1.],
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 0., 1.],
       [0., 1., 0.]], dtype=float32)